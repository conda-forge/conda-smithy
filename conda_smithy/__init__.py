try:
    from conda_smithy._version import __version__
except ImportError:
    __version__ = "0.0.0"
