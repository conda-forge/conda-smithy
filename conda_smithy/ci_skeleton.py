"""Generate skeleton for using conda-forge CI configuration files outside of
a feedstock (i.e. a outside of a forge context). This enables people to run
their CI without having to worry about how to set it up or keep it up-to-date,
by reusing the same tools that conda-forge uses for its infrastructure.
Note that your CI jobs will still execute under your organization, and not be
added to conda-forge's queue.
"""
import os
import sys

from jinja2 import Environment, FileSystemLoader

from .configure_feedstock import make_jinja_env, conda_forge_content


def _render_template(template_file, env, forge_dir, config):
    """Renders the template"""
    template = env.get_template(
        os.path.basename(template_file) + ".ci-skel.tmpl"
    )
    target_fname = os.path.join(forge_dir, template_file)
    print("Generating " + target_fname, file=sys.stderr)
    new_file_contents = template.render(**config)
    os.makedirs(os.path.dirname(target_fname), exist_ok=True)
    with open(target_fname, "w") as fh:
        fh.write(new_file_contents)


def _insert_into_gitignore(
    feedstock_directory=".",
    prefix="# conda smithy ci-skeleton start\n",
    suffix="# conda smithy ci-skeleton end\n",
):
    """Places gitignore contents into gitignore."""
    # get current contents
    fname = os.path.join(feedstock_directory, ".gitignore")
    print("Updating " + fname)
    if os.path.isfile(fname):
        with open(fname, "r") as f:
            s = f.read()
        before, _, s = s.partition(prefix)
        _, _, after = s.partition(suffix)
    else:
        before = after = ""
        dname = os.path.dirname(fname)
        if dname:
            os.makedirs(dname, exist_ok=True)
    # get new values
    gi = os.path.join(conda_forge_content, "feedstock_content", ".gitignore")
    with open(gi, "r") as f:
        s = f.read()
    new = prefix + s + suffix
    # write out the file
    with open(fname, "w") as f:
        f.write(before + new + after)
    return fname


def generate(
    package_name="pkg", feedstock_directory=".", recipe_directory="recipe"
):
    """Generates the CI skeleton."""
    forge_dir = os.path.abspath(feedstock_directory)
    env = make_jinja_env(forge_dir)
    config = dict(
        package_name=package_name,
        feedstock_directory=feedstock_directory,
        recipe_directory=recipe_directory,
    )
    # render templates
    _render_template("conda-forge.yml", env, forge_dir, config)
    _render_template(
        os.path.join(recipe_directory, "meta.yaml"), env, forge_dir, config
    )
    # update files which may exist with other content
    _insert_into_gitignore(feedstock_directory=feedstock_directory)
