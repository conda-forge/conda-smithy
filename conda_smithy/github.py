from __future__ import absolute_import, print_function

import os
import random
from random import choice

import git
from git import Repo

import github
from github import Github
from github.GithubException import GithubException
from github.Organization import Organization
from github.Team import Team

from . import configure_feedstock


def gh_token():
    try:
        with open(os.path.expanduser('~/.conda-smithy/github.token'), 'r') as fh:
            token = fh.read().strip()
    except IOError:
        msg = ('No github token. Go to https://github.com/settings/tokens/new and generate\n'
               'a token with repo access. Put it in ~/.conda-smithy/github.token')
        raise RuntimeError(msg)
    return token


def create_team(org, name, description, repo_names=[]):
    # PyGithub creates secret teams, and has no way of turning that off! :(
    post_parameters = {
        "name": name,
        "description": description,
        "privacy": "closed",
        "permission": "push",
        "repo_names": repo_names
    }
    headers, data = org._requester.requestJsonAndCheck(
        "POST",
        org.url + "/teams",
        input=post_parameters
    )
    return Team(org._requester, headers, data, completed=True)


def add_membership(team, member):
    headers, data = team._requester.requestJsonAndCheck(
        "PUT",
        team.url + "/memberships/" + member
    )
    return (headers, data)


def remove_membership(team, member):
    headers, data = team._requester.requestJsonAndCheck(
        "DELETE",
        team.url + "/memberships/" + member
    )
    return (headers, data)


def has_in_members(team, member):
    status, headers, data = team._requester.requestJson(
        "GET",
        team.url + "/members/" + member
    )
    return status == 204


def get_cached_team(org, team_name, description=""):
    cached_file = os.path.expanduser('~/.conda-smithy/{}-{}-team'.format(org.login, team_name))
    try:
        with open(cached_file, 'r') as fh:
            team_id = int(fh.read().strip())
            return org.get_team(team_id)
    except IOError:
        pass

    team = next((team for team in org.get_teams() if team.name == team_name), None)
    if not team:
        team = create_team(org, team_name, description, [])

    with open(cached_file, 'w') as fh:
        fh.write(str(team.id))

    return team


def create_github_repo(args):
    token = gh_token()
    meta = configure_feedstock.meta_of_feedstock(args.feedstock_directory)

    gh = Github(token)
    user_or_org = None
    if args.user is not None:
        pass
        # User has been defined, and organization has not.
        user_or_org = gh.get_user()
    else:
        # Use the organization provided.
        user_or_org = gh.get_organization(args.organization)

    repo_name = '{}-feedstock'.format(meta.name())
    try:
        gh_repo = user_or_org.create_repo(repo_name, has_wiki=False,
                                          description='A conda-smithy repository for {}.'.format(meta.name()))
        print('Created {} on github'.format(gh_repo.full_name))
    except GithubException as gh_except:
        if gh_except.data.get('errors', [{}])[0].get('message', '') != u'name already exists on this account':
            raise
        gh_repo = user_or_org.get_repo(repo_name)
        print('Github repository already exists.')

    # Now add this new repo as a remote on the local clone.
    repo = Repo(args.feedstock_directory)
    remote_name = args.remote_name.strip()
    if remote_name:
        if remote_name in [remote.name for remote in repo.remotes]:
            existing_remote = repo.remotes[remote_name]
            if existing_remote.url != gh_repo.ssh_url:
                print("Remote {} already exists, and doesn't point to {} "
                      "(it points to {}).".format(remote_name, gh_repo.ssh_url, existing_remote.url))
        else:
            repo.create_remote(remote_name, gh_repo.ssh_url)

    if args.add_teams:
        if isinstance(user_or_org, Organization):
            configure_github_team(meta, gh_repo, user_or_org)


def configure_github_team(meta, gh_repo, org):

    # Add a team for this repo and add the maintainers to it.
    superlative = [
        'awesome', 'slick', 'formidable', 'awe-inspiring',
        'breathtaking', 'magnificent', 'wonderous', 'stunning',
        'astonishing', 'superb', 'splendid', 'impressive',
        'unbeatable', 'excellent', 'top', 'outstanding', 'exalted',
        'standout', 'smashing'
    ]

    maintainers = set(
        meta.meta.get('extra', {}).get('recipe-maintainers', [])
    )
    maintainers = set(maintainer.lower() for maintainer in maintainers)
    team_name = meta.name()
    # Try to get team or create it if it doesn't exist.
    team = next((team for team in gh_repo.get_teams() if team.name == team_name), None)
    current_maintainers = []
    if not team:
        team = create_team(
            org,
            team_name,
            'The {} {} contributors!'.format(
                choice(superlative), team_name
            )
        )
        team.add_to_repos(gh_repo)
    else:
        current_maintainers = team.get_members()


    # Add only the new maintainers to the team.
    current_maintainers_handles = set([
        e.login.lower() for e in current_maintainers
    ])
    for new_maintainer in maintainers - current_maintainers_handles:
        add_membership(team, new_maintainer)

    # Mention any maintainers that need to be removed (unlikely here).
    for old_maintainer in current_maintainers_handles - maintainers:
        print(
            "AN OLD MEMBER ({}) NEEDS TO BE REMOVED FROM {}".format(
                old_maintainer, repo_name
            )
        )

    # Get the all-members team
    team_name = 'all-members'
    description = "All of the awesome {} contributors!".format(org.name)
    all_members_team = get_cached_team(org, team_name, description)
    new_conda_forge_members = set()

    # Add new members to all-members
    for new_member in maintainers - current_maintainers_handles:
        if not has_in_members(all_members_team, new_member):
            print(
                "Adding a new member ({}) to {}. Welcome! :)".format(
                    new_member, org.name
                )
            )
            add_membership(all_members_team, new_member)
            new_conda_forge_members.add(new_member)

    return maintainers, current_maintainers_handles, new_conda_forge_members
