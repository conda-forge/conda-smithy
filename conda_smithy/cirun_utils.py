"""
See http://py.cirun.io/api.html for cirun client docs
"""

from functools import lru_cache
from typing import Any, Dict, List, Optional

from cirun import Cirun

from conda_smithy.github import Github, gh_token


@lru_cache
def get_cirun_installation_id(owner: str) -> int:
    # This ID needs a token with admin: org privileges.
    # Hard-code instead for easier use.
    if owner == "conda-forge":
        return 18453316
    else:
        gh = Github(gh_token())
        user = gh.get_user()
        user_or_org: Any
        if user.login == owner:
            user_or_org = user
        else:
            user_or_org = gh.get_organization(owner)
        for inst in user_or_org.get_installations():
            if inst.raw_data["app_slug"] == "cirun-application":
                return inst.app_id
        raise ValueError(f"cirun not found for owner {owner}")


def enable_cirun_for_project(owner: str, repo: str) -> Dict[str, Any]:
    """Enable the cirun.io Github Application for a particular repository."""
    print(f"Enabling cirun for {owner}/{repo} ...")
    cirun = _get_cirun_client()
    return cirun.set_repo(
        f"{owner}/{repo}", installation_id=get_cirun_installation_id(owner)
    )


def add_repo_to_cirun_resource(
    owner: str,
    repo: str,
    resources: List[str],
    teams: Optional[List],
    roles: Optional[List],
    users_from_json: Optional[str] = None,
    cirun_policy_args: Optional[List[str]] = None,
) -> Dict[str, Any]:
    """Grant access to a cirun resource to a particular repository, with a particular policy."""
    cirun = _get_cirun_client()
    policy_args: Optional[Dict[str, Any]] = None
    if cirun_policy_args and "pull_request" in cirun_policy_args:
        policy_args = {"pull_request": True}
    print(
        f"Adding repo {owner}/{repo} to resources {resources} with policy_args: {policy_args}"
    )
    gh = Github(gh_token())
    gh_owner = gh.get_user(owner)
    gh_repo = gh_owner.get_repo(repo)

    # Need to send None instead of an empty list
    teams = teams or None
    roles = roles or None
    if not (teams or roles or users_from_json):
        teams = [team.name for team in gh_repo.get_teams()]

    response = cirun.add_repo_to_resources(
        owner,
        repo,
        resources=resources,
        policy_args=policy_args,
        teams=teams,
        roles=roles,
        users_from_json=users_from_json,
    )
    print(f"response: {response} | {response.json().keys()}")
    return response


def remove_repo_from_cirun_resource(owner: str, repo: str, resource: str):
    """Revoke access to a cirun resource to a particular repository, with a particular policy."""
    cirun = _get_cirun_client()
    print(f"Removing repo {owner}/{repo} from resource {resource}.")
    response = cirun.remove_repo_from_resources(owner, repo, [resource])
    print(f"response: {response} | {response.json().keys()}")
    return response


@lru_cache
def _get_cirun_client() -> Cirun:
    try:
        return Cirun()
    except KeyError:
        raise RuntimeError(
            "You must have CIRUN_API_KEY defined to do Cirun CI registration. "
            "This requirement can be overriden by specifying `--without-cirun`"
        )
