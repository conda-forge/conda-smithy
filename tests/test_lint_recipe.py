#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import print_function, unicode_literals
from collections import OrderedDict
from contextlib import contextmanager
import io
import os
import shutil
import subprocess
import tempfile
import textwrap
import unittest
import warnings
import github

import conda_smithy.lint_recipe as linter
_thisdir = os.path.abspath(os.path.dirname(__file__))


def is_gh_token_set():
    return 'GH_TOKEN' in os.environ


@contextmanager
def tmp_directory():
    tmp_dir = tempfile.mkdtemp('recipe_')
    yield tmp_dir
    shutil.rmtree(tmp_dir)


class Test_linter(unittest.TestCase):
    def test_bad_top_level(self):
        meta = OrderedDict([['package', {}],
                            ['build', {}],
                            ['sources', {}]])
        lints = linter.lintify(meta)
        expected_msg = ("The top level meta key sources is unexpected")
        self.assertIn(expected_msg, lints)

    def test_bad_order(self):
        meta = OrderedDict([['package', {}],
                            ['build', {}],
                            ['source', {}]])
        lints = linter.lintify(meta)
        expected_msg = ("The top level meta keys are in an unexpected "
                        "order. Expecting ['package', 'source', 'build'].")
        self.assertIn(expected_msg, lints)

    def test_missing_about_license_and_summary(self):
        meta = {'about': {'home': 'a URL'}}
        lints = linter.lintify(meta)
        expected_message = "The license item is expected in the about section."
        self.assertIn(expected_message, lints)

        expected_message = "The summary item is expected in the about section."
        self.assertIn(expected_message, lints)

    def test_bad_about_license(self):
        meta = {'about': {'home': 'a URL',
                          'summary': 'A test summary',
                          'license': 'unknown'}}
        lints = linter.lintify(meta)
        expected_message = "The recipe license cannot be unknown."
        self.assertIn(expected_message, lints)

    def test_bad_about_license_family(self):
        meta = {'about': {'home': 'a URL',
                          'summary': 'A test summary',
                          'license': 'BSD 3-clause',
                          'license_family': 'BSD3'}}
        lints = linter.lintify(meta)
        expected = "about/license_family 'BSD3' not allowed"
        self.assertTrue(any(lint.startswith(expected) for lint in lints))

    def test_missing_about_home(self):
        meta = {'about': {'license': 'BSD',
                          'summary': 'A test summary'}}
        lints = linter.lintify(meta)
        expected_message = "The home item is expected in the about section."
        self.assertIn(expected_message, lints)

    def test_missing_about_home_empty(self):
        meta = {'about': {'home': '',
                          'summary': '',
                          'license': ''}}
        lints = linter.lintify(meta)
        expected_message = "The home item is expected in the about section."
        self.assertIn(expected_message, lints)

        expected_message = "The license item is expected in the about section."
        self.assertIn(expected_message, lints)

        expected_message = "The summary item is expected in the about section."
        self.assertIn(expected_message, lints)

    def test_maintainers_section(self):
        expected_message = ('The recipe could do with some maintainers listed '
                            'in the `extra/recipe-maintainers` section.')

        lints = linter.lintify({'extra': {'recipe-maintainers': []}})
        self.assertIn(expected_message, lints)

        # No extra section at all.
        lints = linter.lintify({})
        self.assertIn(expected_message, lints)

        lints = linter.lintify({'extra': {'recipe-maintainers': ['a']}})
        self.assertNotIn(expected_message, lints)

        expected_message = ('The "extra" section was expected to be a '
                            'dictionary, but got a list.')
        lints = linter.lintify({'extra': ['recipe-maintainers']})
        self.assertIn(expected_message, lints)

        lints = linter.lintify({'extra': {'recipe-maintainers': 'Luke'}})
        expected_message = ('Recipe maintainers should be a json list.')
        self.assertIn(expected_message, lints)

    def test_test_section(self):
        expected_message = 'The recipe must have some tests.'

        lints = linter.lintify({})
        self.assertIn(expected_message, lints)

        lints = linter.lintify({'test': {'files': 'foo'}})
        self.assertIn(expected_message, lints)

        lints = linter.lintify({'test': {'imports': 'sys'}})
        self.assertNotIn(expected_message, lints)

    def test_test_section_with_recipe(self):
        # If we have a run_test.py file, we shouldn't need to provide
        # other tests.

        expected_message = 'The recipe must have some tests.'

        with tmp_directory() as recipe_dir:
            lints = linter.lintify({}, recipe_dir)
            self.assertIn(expected_message, lints)

            with io.open(os.path.join(recipe_dir, 'run_test.py'), 'w') as fh:
                fh.write('# foo')
            lints = linter.lintify({}, recipe_dir)
            self.assertNotIn(expected_message, lints)

    def test_selectors(self):
        expected_message = ('Selectors are suggested to take a '
                         '``<two spaces>#<one space>[<expression>]`` form.'
                         ' See lines {}'.format([3]))

        with tmp_directory() as recipe_dir:
            def assert_selector(selector, is_good=True):
                with io.open(os.path.join(recipe_dir, 'meta.yaml'), 'w') as fh:
                    fh.write("""
                            package:
                               name: foo_py2  # [py2k]
                               {}
                             """.format(selector))
                lints = linter.lintify({}, recipe_dir)
                if is_good:
                    message = ("Found lints when there shouldn't have been a "
                               "lint for '{}'.".format(selector))
                else:
                    message = ("Expecting lints for '{}', but didn't get any."
                               "".format(selector))
                self.assertEqual(not is_good,
                                 any(lint.startswith(expected_message)
                                     for lint in lints),
                                 message)

            assert_selector("name: foo_py3      # [py3k]")
            assert_selector("name: foo_py3  [py3k]", is_good=False)
            assert_selector("name: foo_py3  #[py3k]", is_good=False)
            assert_selector("name: foo_py3 # [py3k]", is_good=False)

    def test_noarch_selectors(self):
        expected_start = "`noarch` packages can't have selectors."

        with tmp_directory() as recipe_dir:
            def assert_noarch_selector(meta_string, is_good=False):
                with io.open(os.path.join(recipe_dir, 'meta.yaml'), 'w') as fh:
                    fh.write(meta_string)
                lints = linter.main(recipe_dir)
                if is_good:
                    message = ("Found lints when there shouldn't have "
                               "been a lint for '{}'."
                              ).format(meta_string)
                else:
                    message = ("Expected lints for '{}', but didn't "
                               "get any.").format(meta_string)
                self.assertEqual(not is_good,
                                 any(lint.startswith(expected_start)
                                     for lint in lints),
                                 message)

            assert_noarch_selector("""
                            build:
                              noarch: python
                              skip: true  # [py2k]
                            """)
            assert_noarch_selector("""
                            build:
                              noarch: generic
                              skip: true  # [win]
                            """)
            assert_noarch_selector("""
                            build:
                              noarch: python
                              skip: true  #
                            """, is_good=True)
            assert_noarch_selector("""
                            build:
                              noarch: python
                              script:
                                - echo "hello" # [unix]
                                - echo "hello" # [win]
                            """, is_good=True)
            assert_noarch_selector("""
                            build:
                              noarch: python
                              script:
                                - echo "hello" # [unix]
                                - echo "hello" # [win]
                              requirements:
                                build:
                                  - python
                            """, is_good=True)
            assert_noarch_selector("""
                            build:
                              noarch: python
                              script:
                                - echo "hello" # [unix]
                                - echo "hello" # [win]
                              requirements:
                                build:
                                  - python
                              tests:
                                commands:
                                  - cp asd qwe  # [unix]
                            """, is_good=True)
            assert_noarch_selector("""
                            build:
                              noarch: python
                              script:
                                - echo "hello" # [unix]
                                - echo "hello" # [win]
                              requirements:
                                build:
                                  - python
                                  - enum34     # [py2k]
                              tests:
                                commands:
                                  - cp asd qwe  # [unix]
                            """)

    def test_jinja_os_environ(self):
        # Test that we can use os.environ in a recipe. We don't care about
        # the results here.
        with tmp_directory() as recipe_dir:
            with io.open(os.path.join(recipe_dir, 'meta.yaml'), 'w') as fh:
                fh.write("""
                        {% set version = os.environ.get('WIBBLE') %}
                        package:
                           name: foo
                           version: {{ version }}
                         """)
            lints = linter.main(recipe_dir)

    def test_missing_build_number(self):
        expected_message = "The recipe must have a `build/number` section."

        meta = {'build': {'skip': 'True',
                          'script': 'python setup.py install',
                          'number': 0}}
        lints = linter.lintify(meta)
        self.assertNotIn(expected_message, lints)

        meta = {'build': {'skip': 'True',
                          'script': 'python setup.py install'}}
        lints = linter.lintify(meta)
        self.assertIn(expected_message, lints)

    def test_bad_requirements_order(self):
        expected_message = ("The `requirements/build` section should be "
                            "defined before the `requirements/run` section.")

        meta = {'requirements': OrderedDict([['run', 'a'],
                                             ['build', 'a']])}
        lints = linter.lintify(meta)
        self.assertIn(expected_message, lints)

        meta = {'requirements': OrderedDict([['run', 'a'],
                                             ['invalid', 'a'],
                                             ['build', 'a']])}
        lints = linter.lintify(meta)
        self.assertIn(expected_message, lints)

        meta = {'requirements': OrderedDict([['build', 'a'],
                                             ['run', 'a']])}
        lints = linter.lintify(meta)
        self.assertNotIn(expected_message, lints)

    def test_no_sha_with_dl(self):
        expected_message = ("When defining a source/url please add a sha256, "
                            "sha1 or md5 checksum (sha256 preferably).")
        meta = {'source': {'url': None}}
        self.assertIn(expected_message, linter.lintify(meta))

        meta = {'source': {'url': None, 'sha1': None}}
        self.assertNotIn(expected_message, linter.lintify(meta))

        meta = {'source': {'url': None, 'sha256': None}}
        self.assertNotIn(expected_message, linter.lintify(meta))

        meta = {'source': {'url': None, 'md5': None}}
        self.assertNotIn(expected_message, linter.lintify(meta))

    def test_redundant_license(self):
        meta = {'about': {'home': 'a URL',
                          'summary': 'A test summary',
                          'license': 'MIT License'}}
        lints = linter.lintify(meta)
        expected_message = ('The recipe `license` should not include '
                            'the word "License".')
        self.assertIn(expected_message, lints)

    def test_recipe_name(self):
        meta = {'package': {'name': 'mp++'}}
        lints = linter.lintify(meta)
        expected_message = ('Recipe name has invalid characters. only lowercase alpha, '
                            'numeric, underscores, hyphens and dots allowed')
        self.assertIn(expected_message, lints)

    def test_end_empty_line(self):
        bad_contents = [
            # No empty lines at the end of the file
            'extra:\n  recipe-maintainers:\n    - goanpeca',
            'extra:\r  recipe-maintainers:\r    - goanpeca',
            'extra:\r\n  recipe-maintainers:\r\n    - goanpeca',
            # Two empty lines at the end of the file
            'extra:\n  recipe-maintainers:\n    - goanpeca\n\n',
            'extra:\r  recipe-maintainers:\r    - goanpeca\r\r',
            'extra:\r\n  recipe-maintainers:\r\n    - goanpeca\r\n\r\n',
            # Three empty lines at the end of the file
            'extra:\n  recipe-maintainers:\n    - goanpeca\n\n\n',
            'extra:\r  recipe-maintainers:\r    - goanpeca\r\r\r',
            'extra:\r\n  recipe-maintainers:\r\n    - goanpeca\r\n\r\n\r\n',
        ]
        # Exactly one empty line at the end of the file
        valid_content = 'extra:\n  recipe-maintainers:\n    - goanpeca\n'

        for content, lines in zip(bad_contents + [valid_content],
                                  [0, 0, 0, 2, 2, 2, 3, 3, 3, 1]):
            with tmp_directory() as recipe_dir:
                with io.open(os.path.join(recipe_dir, 'meta.yaml'), 'w') as f:
                    f.write(content)
                lints = linter.lintify({}, recipe_dir=recipe_dir)
                if lines > 1:
                    expected_message = ('There are {} too many lines.  '
                                        'There should be one empty line '
                                        'at the end of the '
                                        'file.'.format(lines - 1))
                else:
                    expected_message = ('There are too few lines.  '
                                        'There should be one empty line at'
                                        ' the end of the file.')
                if content == valid_content:
                    self.assertNotIn(expected_message, lints)
                else:
                    self.assertIn(expected_message, lints)

    def test_cb3_jinja2_functions(self):
        lints = linter.main(os.path.join(_thisdir, 'recipes', 'cb3_jinja2_functions', 'recipe'))
        assert not lints

    @unittest.skipUnless(is_gh_token_set(), "GH_TOKEN not set")
    def test_maintainer_exists(self):
        lints = linter.lintify({'extra': {'recipe-maintainers': ['support']}}, conda_forge=True)
        expected_message = ('Recipe maintainer "support" does not exist')
        self.assertIn(expected_message, lints)

        lints = linter.lintify({'extra': {'recipe-maintainers': ['isuruf']}}, conda_forge=True)
        expected_message = ('Recipe maintainer "isuruf" does not exist')
        self.assertNotIn(expected_message, lints)

        expected_message = 'Feedstock with the same name exists in conda-forge'
        # Check that feedstock exists if staged_recipes
        lints = linter.lintify({'package': {'name': 'python'}}, recipe_dir='python', conda_forge=True)
        self.assertIn(expected_message, lints)
        lints = linter.lintify({'package': {'name': 'python'}}, recipe_dir='python', conda_forge=False)
        self.assertNotIn(expected_message, lints)
        # No lint if in a feedstock
        lints = linter.lintify({'package': {'name': 'python'}}, recipe_dir='recipe', conda_forge=True)
        self.assertNotIn(expected_message, lints)
        lints = linter.lintify({'package': {'name': 'python'}}, recipe_dir='recipe', conda_forge=False)
        self.assertNotIn(expected_message, lints)

        # Make sure there's no feedstock named python1 before proceeding
        gh = github.Github(os.environ['GH_TOKEN'])
        cf = gh.get_user('conda-forge')
        try:
            cf.get_repo('python1-feedstock')
            feedstock_exists = True
        except github.UnknownObjectException as e:
            feedstock_exists = False

        if feedstock_exists:
            warnings.warn("There's a feedstock named python1, but tests assume that there isn't")
        else:
            lints = linter.lintify({'package': {'name': 'python1'}}, recipe_dir="python", conda_forge=True)
            self.assertNotIn(expected_message, lints)


    def test_bad_subheader(self):
        expected_message = 'The {} section contained an unexpected ' \
                           'subsection name. {} is not a valid subsection' \
                           ' name.'.format('build', 'ski')
        meta = {'build': {'skip': 'True',
                          'script': 'python setup.py install',
                          'number': 0}}
        lints = linter.lintify(meta)
        self.assertNotIn(expected_message, lints)

        meta = {'build': {'ski': 'True',
                          'script': 'python setup.py install',
                          'number': 0}}
        lints = linter.lintify(meta)
        self.assertIn(expected_message, lints)

    def test_outputs(self):
        meta = OrderedDict([['outputs', [{'name': 'asd'}]]])
        lints = linter.lintify(meta)

    def test_version(self):
        meta = {'package': {'name': 'python',
                            'version': '3.6.4'}}
        expected_message = "Package version 3.6.4 doesn't match conda spec"
        lints = linter.lintify(meta)
        self.assertNotIn(expected_message, lints)

        meta = {'package': {'name': 'python',
                            'version': '2.0.0~alpha0'}}
        expected_message = "Package version 2.0.0~alpha0 doesn't match conda spec"
        lints = linter.lintify(meta)
        self.assertIn(expected_message, lints)


class TestCLI_recipe_lint(unittest.TestCase):
    def test_cli_fail(self):
        with tmp_directory() as recipe_dir:
            with io.open(os.path.join(recipe_dir, 'meta.yaml'), 'w') as fh:
                fh.write(textwrap.dedent("""
                    package:
                        name: 'test_package'
                    build: []
                    requirements: []
                    """))
            child = subprocess.Popen(['conda-smithy', 'recipe-lint',
                                      recipe_dir],
                                     stdout=subprocess.PIPE)
            out, _ = child.communicate()
            self.assertEqual(child.returncode, 1, out)

    def test_cli_success(self):
        with tmp_directory() as recipe_dir:
            with io.open(os.path.join(recipe_dir, 'meta.yaml'), 'w') as fh:
                fh.write(textwrap.dedent("""
                    package:
                        name: 'test_package'
                    build:
                        number: 0
                    test:
                        imports:
                            - foo
                    about:
                        home: something
                        license: something else
                        summary: a test recipe
                    extra:
                        recipe-maintainers:
                            - a
                            - b
                    """))
            child = subprocess.Popen(['conda-smithy', 'recipe-lint',
                                      recipe_dir],
                                     stdout=subprocess.PIPE)
            out, _ = child.communicate()
            self.assertEqual(child.returncode, 0, out)

    def test_cli_environ(self):
        with tmp_directory() as recipe_dir:
            with io.open(os.path.join(recipe_dir, 'meta.yaml'), 'w') as fh:
                fh.write(textwrap.dedent("""
                    package:
                        name: 'test_package'
                    build:
                        number: 0
                    test:
                        requires:
                            - python {{ environ['PY_VER'] + '*' }}  # [win]
                        imports:
                            - foo
                    about:
                        home: something
                        license: something else
                        summary: a test recipe
                    extra:
                        recipe-maintainers:
                            - a
                            - b
                    """))
            child = subprocess.Popen(['conda-smithy', 'recipe-lint',
                                      recipe_dir],
                                     stdout=subprocess.PIPE)
            out, _ = child.communicate()
            self.assertEqual(child.returncode, 0, out)

    def test_unicode(self):
        """
        Tests that unicode does not confuse the linter.
        """
        with tmp_directory() as recipe_dir:
            with io.open(os.path.join(recipe_dir, 'meta.yaml'), 'wt', encoding='utf-8') as fh:
                fh.write("""
                    package:
                        name: 'test_package'
                    build:
                        number: 0
                    about:
                        home: something
                        license: something else
                        summary: αβɣ
                        description: moɿɘ uniɔobɘ!
                         """)
            # Just run it and make sure it does not raise.
            linter.main(recipe_dir)


if __name__ == '__main__':
    unittest.main()
