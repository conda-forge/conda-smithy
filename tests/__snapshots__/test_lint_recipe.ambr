# serializer version: 1
# name: test_osx_hint[run]
  list([
    'You\'re setting a constraint on the `__osx` virtual package directly; this should now be done by adding a build dependence on `{{ stdlib("c") }}`, and overriding `c_stdlib_version` in `recipe/conda_build_config.yaml` for the respective platform as necessary. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.',
  ])
# ---
# name: test_osx_hint[run_constrained]
  list([
    'You\'re setting a constraint on the `__osx` virtual package directly; this should now be done by adding a build dependence on `{{ stdlib("c") }}`, and overriding `c_stdlib_version` in `recipe/conda_build_config.yaml` for the respective platform as necessary. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.',
  ])
# ---
# name: test_osx_noarch_hint[run]
  list([
  ])
# ---
# name: test_osx_noarch_hint[run_constrained]
  list([
  ])
# ---
# name: test_stdlib_hint[c]
  list([
    'This recipe is using a compiler, which now requires adding a build dependence on `{{ stdlib("c") }}` as well. Note that this rule applies to each output of the recipe using a compiler. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.',
  ])
# ---
# name: test_stdlib_hint[cxx]
  list([
    'This recipe is using a compiler, which now requires adding a build dependence on `{{ stdlib("c") }}` as well. Note that this rule applies to each output of the recipe using a compiler. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.',
  ])
# ---
# name: test_stdlib_hint[fortran]
  list([
    'This recipe is using a compiler, which now requires adding a build dependence on `{{ stdlib("c") }}` as well. Note that this rule applies to each output of the recipe using a compiler. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.',
  ])
# ---
# name: test_stdlib_hint[m2w64_c]
  list([
    'This recipe is using a compiler, which now requires adding a build dependence on `{{ stdlib("c") }}` as well. Note that this rule applies to each output of the recipe using a compiler. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.',
  ])
# ---
# name: test_stdlib_hint[m2w64_cxx]
  list([
    'This recipe is using a compiler, which now requires adding a build dependence on `{{ stdlib("c") }}` as well. Note that this rule applies to each output of the recipe using a compiler. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.',
  ])
# ---
# name: test_stdlib_hint[m2w64_fortran]
  list([
    'This recipe is using a compiler, which now requires adding a build dependence on `{{ stdlib("c") }}` as well. Note that this rule applies to each output of the recipe using a compiler. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.',
  ])
# ---
# name: test_stdlib_hint[rust]
  list([
    'This recipe is using a compiler, which now requires adding a build dependence on `{{ stdlib("c") }}` as well. Note that this rule applies to each output of the recipe using a compiler. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.',
  ])
# ---
# name: test_stdlib_hints_multi_output
  list([
    'This recipe is using a compiler, which now requires adding a build dependence on `{{ stdlib("c") }}` as well. Note that this rule applies to each output of the recipe using a compiler. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.',
    'You\'re setting a requirement on sysroot_linux-<arch> directly; this should now be done by adding a build dependence on `{{ stdlib("c") }}`, and overriding `c_stdlib_version` in `recipe/conda_build_config.yaml` for the respective platform as necessary. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.',
    'You\'re setting a constraint on the `__osx` virtual package directly; this should now be done by adding a build dependence on `{{ stdlib("c") }}`, and overriding `c_stdlib_version` in `recipe/conda_build_config.yaml` for the respective platform as necessary. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.',
  ])
# ---
# name: test_sysroot_hint
  list([
    'You\'re setting a requirement on sysroot_linux-<arch> directly; this should now be done by adding a build dependence on `{{ stdlib("c") }}`, and overriding `c_stdlib_version` in `recipe/conda_build_config.yaml` for the respective platform as necessary. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.',
  ])
# ---
